components/script/dom/servoparser/async_html.rs

```diff
@@ -20,218 +21,249 @@ use dom::htmltemplateelement::HTMLTemplateElement;
-                node
-            });
-            let tb = TreeBuilder::new_for_fragment(
-                sink,
-                ctxt_parse_node,
-                form_parse_node,
-                options);
-
-            let tok_options = TokenizerOpts {
-                initial_state: Some(tb.tokenizer_state_for_context_elem()),
-                .. Default::default()
-            };
-
-            HtmlTokenizer::new(tb, tok_options)
-        } else {
-            HtmlTokenizer::new(TreeBuilder::new(sink, options), Default::default())
-        };
-
-        Tokenizer {
-            inner: inner,
-        }
-    }
-
-    pub fn feed(&mut self, input: &mut BufferQueue) -> Result<(), Root<HTMLScriptElement>> {
-        match self.inner.feed(input) {
-            TokenizerResult::Done => Ok(()),
-            TokenizerResult::Script(script) => {
-                let nodes = &self.inner.sink.sink.nodes;
-                let script = nodes.get(&script.id).unwrap();
-                Err(Root::from_ref(script.downcast().unwrap()))
-            },
-        }
-    }
-
-    pub fn end(&mut self) {
-        self.inner.end();
-    }
-
-    pub fn url(&self) -> &ServoUrl {
-        &self.inner.sink.sink.base_url
-    }
-
-    pub fn set_plaintext_state(&mut self) {
-        self.inner.set_plaintext_state();
-    }
-}
-
-#[allow(unsafe_code)]
-unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<ParseNode, Sink>> {
-    unsafe fn trace(&self, trc: *mut JSTracer) {
-        struct Tracer(*mut JSTracer);
-        let tracer = Tracer(trc);
-
-        impl HtmlTracer for Tracer {
-            type Handle = ParseNode;
-            #[allow(unrooted_must_root)]
-            fn trace_handle(&self, node: &ParseNode) {
-                unsafe { node.trace(self.0); }
-            }
-        }
-
-        let tree_builder = &self.sink;
-        tree_builder.trace_handles(&tracer);
-        tree_builder.sink.trace(trc);
-    }
-}
-
 type ParseNodeId = usize;
 
-#[derive(JSTraceable, Clone, HeapSizeOf)]
+#[derive(Clone, HeapSizeOf, JSTraceable)]
 pub struct ParseNode {
     id: ParseNodeId,
     qual_name: Option<QualName>,
 }
 
-#[derive(JSTraceable, HeapSizeOf)]
-struct ParseNodeData {
-    contents: Option<ParseNode>,
-    is_integration_point: bool,
+#[derive(HeapSizeOf, JSTraceable)]
+enum NodeOrText {
+    Node(ParseNode),
+    Text(String),
 }
 
-impl Default for ParseNodeData {
-    fn default() -> ParseNodeData {
-        ParseNodeData {
-            contents: None,
-            is_integration_point: false,
-        }
-    }
+#[derive(HeapSizeOf, JSTraceable)]
+struct Attribute {
+    name: QualName,
+    value: String,
 }
 
+#[derive(HeapSizeOf, JSTraceable)]
 enum ParseOperation {
     GetTemplateContents(ParseNodeId, ParseNodeId),
-    CreateElement(ParseNodeId, QualName, Vec<Attribute>),
-    CreateComment(StrTendril, ParseNodeId),
+    CreateElement(ParseNodeId, QualName, Vec<Attribute>, u64),
```

Comment:
Please make these have named fields, as I mentioned in the first PR. What even is that `u64`?